File: .claude/skills/project-index/scripts/scan-structure.js
#!/usr/bin/env node
/**
 * Project Structure Scanner
 * Cross-platform script to generate an AI-friendly project map + directory tree.
 *
 * Usage:
 *   node scan-structure.js [rootPath] [maxDepth] [format]
 *   node scan-structure.js . 4
 *   node scan-structure.js . 4 json
 *   node scan-structure.js . 4 --no-gitignore
 */

const fs = require('fs');
const os = require('os');
const path = require('path');

const BASE_IGNORES = [
  // VCS / OS
  '.git',
  '.DS_Store',
  'Thumbs.db',

  // Dependencies / build outputs
  'node_modules',
  'dist',
  'build',
  'out',
  'coverage',
  '.next',
  '.nuxt',
  '.turbo',

  // Python virtualenv / caches
  'venv',
  '.venv',
  'env',
  '.tox',
  '__pycache__',
  '.pytest_cache',
  '.mypy_cache',
  '.ruff_cache',

  // Tooling caches / IDE
  '.cache',
  '.idea',
  '.vscode',

  // .NET / Rust typical outputs
  'bin',
  'obj',
  'target',

  // Secrets (files)
  '.env',
  '.env.local',
  '.env.development.local',
  '.env.test.local',
  '.env.production.local',
];

const FILE_CATEGORIES = {
  typescript: ['.ts', '.tsx', '.mts', '.cts'],
  javascript: ['.js', '.jsx', '.mjs', '.cjs'],
  python: ['.py', '.pyw', '.pyi'],
  csharp: ['.cs', '.csx'],
  java: ['.java'],
  go: ['.go'],
  rust: ['.rs'],
  ruby: ['.rb'],
  php: ['.php'],
  styles: ['.css', '.scss', '.sass', '.less', '.styl'],
  markup: ['.html', '.htm', '.vue', '.svelte'],
  config: ['.json', '.yaml', '.yml', '.toml', '.xml', '.ini'],
  markdown: ['.md', '.mdx'],
  other: [],
};

function normalizeRelPath(p) {
  return p.split(path.sep).join('/');
}

function parseGitignore(rootPath) {
  const gitignorePath = path.join(rootPath, '.gitignore');
  if (!fs.existsSync(gitignorePath)) return [];

  try {
    const raw = fs.readFileSync(gitignorePath, 'utf8');
    return raw
      .split(/\r?\n/g)
      .map(line => line.trim())
      .filter(line => line && !line.startsWith('#'))
      .filter(line => !line.startsWith('!'))
      .map(line => (line.startsWith('/') ? line.slice(1) : line));
  } catch {
    return [];
  }
}

function globToRegex(pattern) {
  // Minimal glob: "*" and "?" only. Treat "/" as path separator in normalized paths.
  const escaped = pattern.replace(/[.+^${}()|[\]\\]/g, '\\$&');
  const regex = '^' + escaped.replace(/\*/g, '.*').replace(/\?/g, '.') + '$';
  return new RegExp(regex);
}

function describeTopLevelEntry(name, isDir) {
  if (!isDir) {
    const lower = name.toLowerCase();
    if (lower === 'readme.md') return 'Project overview and getting started';
    if (lower === 'agents.md') return 'Agent working conventions / repo rules';
    if (['changelog.md', 'handoff.md', 'contributing.md', 'license', 'license.md'].includes(lower)) {
      return 'Project process / legal / handoff documentation';
    }
    if (['package.json', 'pnpm-lock.yaml', 'yarn.lock', 'package-lock.json'].includes(lower)) {
      return 'JavaScript/TypeScript dependencies and scripts';
    }
    if (['pyproject.toml', 'requirements.txt', 'poetry.lock', 'pipfile', 'pipfile.lock'].includes(lower)) {
      return 'Python dependencies and tooling';
    }
    if (['.gitignore', '.gitattributes'].includes(lower)) return 'Git configuration';
    return 'Key file (see tree for context)';
  }

  const lower = name.toLowerCase();
  if (lower === 'src') return 'Main application source code';
  if (['app', 'apps'].includes(lower)) return 'Application(s) (often runnable targets)';
  if (['packages', 'pkg'].includes(lower)) return 'Packages (shared modules/libraries)';
  if (['libs', 'lib'].includes(lower)) return 'Shared libraries/utilities';
  if (['services', 'service'].includes(lower)) return 'Backend services / service layer';
  if (['api', 'apis'].includes(lower)) return 'API layer (routes/controllers/handlers)';
  if (['tests', 'test', '__tests__'].includes(lower)) return 'Tests';
  if (['docs', 'documentation'].includes(lower)) return 'Documentation';
  if (['scripts', 'tools'].includes(lower)) return 'Automation scripts / tooling';
  if (['public', 'static', 'assets'].includes(lower)) return 'Static assets';
  if (lower === '.github') return 'GitHub configuration (CI, templates, workflows)';
  if (lower === '.claude') return 'AI tooling / agent workflows (project-local)';
  if (lower === '.codex') return 'Codex CLI skills / configuration (project-local)';
  if (['infra', 'infrastructure', 'deploy', 'ops', 'k8s', 'terraform'].includes(lower)) {
    return 'Infrastructure / deployment';
  }
  return 'Project-specific directory (see tree)';
}

class ProjectScanner {
  constructor(rootPath, maxDepth = 4, respectGitignore = true) {
    this.rootPath = path.resolve(rootPath);
    this.maxDepth = maxDepth;
    this.respectGitignore = respectGitignore;
    this.gitignorePatterns = respectGitignore ? parseGitignore(this.rootPath) : [];

    this.stats = {
      totalFiles: 0,
      totalDirs: 0,
      byCategory: {},
      entryPoints: [],
      configFiles: [],
      keyFiles: [],
    };
  }

  shouldIgnore(name, relativePath) {
    if (BASE_IGNORES.includes(name) || BASE_IGNORES.includes(relativePath)) return true;

    const relNorm = normalizeRelPath(relativePath);
    const parts = relNorm ? relNorm.split('/') : [];
    if (parts.some(p => BASE_IGNORES.includes(p))) return true;

    // Apply gitignore patterns (simplified)
    for (const rawPattern of this.gitignorePatterns) {
      const p = normalizeRelPath(rawPattern);

      if (p.endsWith('/')) {
        const dir = p.slice(0, -1);
        if (relNorm === dir || relNorm.startsWith(dir + '/')) return true;
        continue;
      }

      const re = globToRegex(p);
      if (p.includes('/')) {
        if (re.test(relNorm)) return true;
      } else {
        if (re.test(name)) return true;
      }
    }

    return false;
  }

  categorizeFile(fileName) {
    const ext = path.extname(fileName).toLowerCase();
    for (const [category, exts] of Object.entries(FILE_CATEGORIES)) {
      if (exts.includes(ext)) return category;
    }
    return 'other';
  }

  isEntryPoint(fileName) {
    return [
      /^(main|index|app|server|entry)\.(ts|tsx|js|jsx|py|go|rs)$/i,
      /^(Program|Startup)\.(cs)$/i,
    ].some(p => p.test(fileName));
  }

  isConfigFile(fileName) {
    return [
      /^package\.json$/i,
      /^tsconfig.*\.json$/i,
      /^vite\.config\./i,
      /^next\.config\./i,
      /^webpack\.config\./i,
      /^\.eslintrc/i,
      /^\.prettierrc/i,
      /^tailwind\.config\./i,
      /^docker-compose/i,
      /^Dockerfile$/i,
      /^Makefile$/i,
      /^\.env\.example$/i,
      /^requirements\.txt$/i,
      /^pyproject\.toml$/i,
      /^Cargo\.toml$/i,
      /^go\.mod$/i,
      /^.*\.(csproj|sln)$/i,
    ].some(p => p.test(fileName));
  }

  isKeyFile(fileName) {
    return [
      /^README/i,
      /^CHANGELOG/i,
      /^CONTRIBUTING/i,
      /^LICENSE/i,
      /^AGENTS\.md$/i,
      /^HANDOFF\.md$/i,
    ].some(p => p.test(fileName));
  }

  getTopLevelMap() {
    const items = [];
    let entries = [];
    try {
      entries = fs.readdirSync(this.rootPath, { withFileTypes: true });
    } catch {
      return items;
    }

    entries.sort((a, b) => {
      if (a.isDirectory() && !b.isDirectory()) return -1;
      if (!a.isDirectory() && b.isDirectory()) return 1;
      return a.name.localeCompare(b.name);
    });

    for (const entry of entries) {
      const rel = entry.name; // top-level
      if (this.shouldIgnore(entry.name, rel)) continue;
      const isDir = entry.isDirectory();
      items.push({
        path: isDir ? `${entry.name}/` : entry.name,
        type: isDir ? 'dir' : 'file',
        purpose: describeTopLevelEntry(entry.name, isDir),
      });
    }

    return items;
  }

  scanDirectory(dirPath, depth = 0, prefix = '') {
    const lines = [];
    if (depth > this.maxDepth) return lines;

    let entries;
    try {
      entries = fs.readdirSync(dirPath, { withFileTypes: true });
    } catch {
      return lines;
    }

    entries.sort((a, b) => {
      if (a.isDirectory() && !b.isDirectory()) return -1;
      if (!a.isDirectory() && b.isDirectory()) return 1;
      return a.name.localeCompare(b.name);
    });

    // Filter ignored entries (check name + relative path)
    const filtered = [];
    for (const entry of entries) {
      const fullPath = path.join(dirPath, entry.name);
      const rel = normalizeRelPath(path.relative(this.rootPath, fullPath));
      if (!this.shouldIgnore(entry.name, rel)) filtered.push({ entry, rel, fullPath });
    }

    filtered.forEach(({ entry, rel, fullPath }, index) => {
      const isLast = index === filtered.length - 1;
      const connector = isLast ? '`-- ' : '|-- ';
      const childPrefix = isLast ? '    ' : '|   ';

      if (entry.isDirectory()) {
        this.stats.totalDirs++;
        lines.push(`${prefix}${connector}${entry.name}/`);
        lines.push(...this.scanDirectory(fullPath, depth + 1, prefix + childPrefix));
        return;
      }

      this.stats.totalFiles++;
      const category = this.categorizeFile(entry.name);
      this.stats.byCategory[category] = (this.stats.byCategory[category] || 0) + 1;

      if (this.isEntryPoint(entry.name)) this.stats.entryPoints.push(rel);
      if (this.isConfigFile(entry.name)) this.stats.configFiles.push(rel);
      if (this.isKeyFile(entry.name)) this.stats.keyFiles.push(rel);

      lines.push(`${prefix}${connector}${entry.name}`);
    });

    return lines;
  }

  scan() {
    const rootName = path.basename(this.rootPath);
    const lines = [`${rootName}/`, ...this.scanDirectory(this.rootPath)];
    return { tree: lines.join('\n'), stats: this.stats, topLevelMap: this.getTopLevelMap() };
  }

  getMainLanguage() {
    const codeCategories = ['typescript', 'javascript', 'python', 'csharp', 'java', 'go', 'rust', 'ruby', 'php'];
    let maxCategory = 'unknown';
    let maxCount = 0;
    for (const cat of codeCategories) {
      const count = this.stats.byCategory[cat] || 0;
      if (count > maxCount) {
        maxCount = count;
        maxCategory = cat;
      }
    }
    return maxCategory;
  }

  generateMarkdown() {
    const result = this.scan();
    const mainLang = this.getMainLanguage();
    const now = new Date().toISOString().replace('T', ' ').slice(0, 16);

    const topLevelTable = result.topLevelMap.length
      ? result.topLevelMap.map(i => `| ${i.path} | ${i.type} | ${i.purpose} |`).join('\n')
      : '| (none) | - | - |';

    const dist = Object.entries(result.stats.byCategory)
      .sort((a, b) => b[1] - a[1])
      .map(([cat, count]) => `| ${cat} | ${count} |`)
      .join('\n') || '| (none) | 0 |';

    return `# Project Structure Index
> Auto-generated by project-index. Last updated: ${now}

## Quick Stats
- **Total files**: ${result.stats.totalFiles}
- **Total directories**: ${result.stats.totalDirs}
- **Main language**: ${mainLang}

## Top-Level Map
| Path | Type | Purpose |
|:---|:---:|:---|
${topLevelTable}

## Directory Tree
\`\`\`
${result.tree}
\`\`\`

## Entry Points
${result.stats.entryPoints.map(f => `- ${f}`).join('\n') || '- (none detected)'}

## Config Files
${result.stats.configFiles.map(f => `- ${f}`).join('\n') || '- (none detected)'}

## Key Files
${result.stats.keyFiles.map(f => `- ${f}`).join('\n') || '- (none detected)'}

## File Distribution
| Category | Count |
|:---|---:|
${dist}
`;
  }

  generateJSON() {
    const result = this.scan();
    return JSON.stringify(
      {
        generated: new Date().toISOString(),
        rootPath: this.rootPath,
        maxDepth: this.maxDepth,
        respectGitignore: this.respectGitignore,
        stats: result.stats,
        mainLanguage: this.getMainLanguage(),
        topLevelMap: result.topLevelMap,
        tree: result.tree,
      },
      null,
      2
    );
  }
}

// CLI execution
if (require.main === module) {
  const args = process.argv.slice(2);
  const rootPath = args[0] || '.';
  const maxDepth = Number.isFinite(Number(args[1])) ? Number(args[1]) : 4;
  const format = args.find(a => a === 'json') ? 'json' : 'markdown';
  const respectGitignore = !args.includes('--no-gitignore');

  const scanner = new ProjectScanner(rootPath, maxDepth, respectGitignore);
  if (format === 'json') console.log(scanner.generateJSON());
  else console.log(scanner.generateMarkdown());
}

module.exports = { ProjectScanner };

File: .claude/skills/project-index/scripts/scan_structure.py
#!/usr/bin/env python3
"""
Project Structure Scanner
Cross-platform script to generate an AI-friendly project map + directory tree.

Usage:
  python scan_structure.py [root_path] [max_depth] [format]
  python scan_structure.py . 4
  python scan_structure.py . 4 json
  python scan_structure.py . 4 --no-gitignore
"""

from __future__ import annotations

import fnmatch
import json
import os
import re
import sys
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional, Set, Tuple

# Goal: keep output focused on "main" project directories and avoid common noise.
BASE_IGNORES: Set[str] = {
    # VCS / OS
    ".git",
    ".DS_Store",
    "Thumbs.db",
    # Dependencies / build outputs
    "node_modules",
    "dist",
    "build",
    "out",
    "coverage",
    ".next",
    ".nuxt",
    ".turbo",
    # Python virtualenv / caches
    "venv",
    ".venv",
    "env",
    ".tox",
    "__pycache__",
    ".pytest_cache",
    ".mypy_cache",
    ".ruff_cache",
    # Tooling caches / IDE
    ".cache",
    ".idea",
    ".vscode",
    # .NET / Rust typical outputs
    "bin",
    "obj",
    "target",
    # Secrets (files)
    ".env",
    ".env.local",
    ".env.development.local",
    ".env.test.local",
    ".env.production.local",
}


FILE_CATEGORIES: Dict[str, Set[str]] = {
    "typescript": {".ts", ".tsx", ".mts", ".cts"},
    "javascript": {".js", ".jsx", ".mjs", ".cjs"},
    "python": {".py", ".pyw", ".pyi"},
    "csharp": {".cs", ".csx"},
    "java": {".java"},
    "go": {".go"},
    "rust": {".rs"},
    "ruby": {".rb"},
    "php": {".php"},
    "styles": {".css", ".scss", ".sass", ".less", ".styl"},
    "markup": {".html", ".htm", ".vue", ".svelte"},
    "config": {".json", ".yaml", ".yml", ".toml", ".xml", ".ini"},
    "markdown": {".md", ".mdx"},
}


ENTRY_PATTERNS = [
    re.compile(r"^(main|index|app|server|entry)\.(ts|tsx|js|jsx|py|go|rs)$", re.I),
    re.compile(r"^(Program|Startup)\.(cs)$", re.I),
]

CONFIG_PATTERNS = [
    re.compile(r"^package\.json$", re.I),
    re.compile(r"^tsconfig.*\.json$", re.I),
    re.compile(r"^vite\.config\.", re.I),
    re.compile(r"^next\.config\.", re.I),
    re.compile(r"^webpack\.config\.", re.I),
    re.compile(r"^\.eslintrc", re.I),
    re.compile(r"^\.prettierrc", re.I),
    re.compile(r"^tailwind\.config\.", re.I),
    re.compile(r"^docker-compose", re.I),
    re.compile(r"^Dockerfile$", re.I),
    re.compile(r"^Makefile$", re.I),
    re.compile(r"^\.env\.example$", re.I),
    re.compile(r"^requirements\.txt$", re.I),
    re.compile(r"^pyproject\.toml$", re.I),
    re.compile(r"^Cargo\.toml$", re.I),
    re.compile(r"^go\.mod$", re.I),
]

KEY_PATTERNS = [
    re.compile(r"^README", re.I),
    re.compile(r"^CHANGELOG", re.I),
    re.compile(r"^CONTRIBUTING", re.I),
    re.compile(r"^LICENSE", re.I),
    re.compile(r"^AGENTS\.md$", re.I),
    re.compile(r"^HANDOFF\.md$", re.I),
]


def parse_gitignore(gitignore_path: Path) -> List[str]:
    patterns: List[str] = []
    if not gitignore_path.exists():
        return patterns

    try:
        for raw in gitignore_path.read_text(encoding="utf-8").splitlines():
            line = raw.strip()
            if not line or line.startswith("#"):
                continue
            if line.startswith("!"):
                continue
            if line.startswith("/"):
                line = line[1:]
            patterns.append(line)
    except Exception as exc:
        print(f"Warning: Could not read .gitignore: {exc}", file=sys.stderr)

    return patterns


def normalize_rel_path(p: Path) -> str:
    return str(p).replace(os.sep, "/")


class ProjectScanner:
    def __init__(self, root_path: str, max_depth: int = 4, respect_gitignore: bool = True):
        self.root_path = Path(root_path).resolve()
        self.max_depth = max_depth

        self.respect_gitignore = respect_gitignore
        self.gitignore_patterns: List[str] = (
            parse_gitignore(self.root_path / ".gitignore") if respect_gitignore else []
        )

        self.stats = {
            "total_files": 0,
            "total_dirs": 0,
            "by_category": {},
            "entry_points": [],
            "config_files": [],
            "key_files": [],
        }

    def should_ignore(self, name: str, rel_path: str) -> bool:
        if name in BASE_IGNORES or rel_path in BASE_IGNORES:
            return True

        rel_path_norm = rel_path.replace(os.sep, "/")
        parts = rel_path_norm.split("/") if rel_path_norm else []

        # Base ignore by segment
        if any(part in BASE_IGNORES for part in parts):
            return True

        for pattern in self.gitignore_patterns:
            p = pattern.replace(os.sep, "/")

            # Directory pattern (ends with "/")
            if p.endswith("/"):
                p_dir = p[:-1]
                if rel_path_norm == p_dir or rel_path_norm.startswith(p_dir + "/"):
                    return True
                continue

            # Path pattern
            if "/" in p:
                if fnmatch.fnmatch(rel_path_norm, p):
                    return True
                continue

            # Name pattern
            if fnmatch.fnmatch(name, p):
                return True

        return False

    def categorize_file(self, file_name: str) -> str:
        ext = Path(file_name).suffix.lower()
        for category, exts in FILE_CATEGORIES.items():
            if ext in exts:
                return category
        return "other"

    def matches_any(self, name: str, patterns: List[re.Pattern]) -> bool:
        return any(p.search(name) for p in patterns)

    def is_entry_point(self, name: str) -> bool:
        return self.matches_any(name, ENTRY_PATTERNS)

    def is_config_file(self, name: str) -> bool:
        return self.matches_any(name, CONFIG_PATTERNS)

    def is_key_file(self, name: str) -> bool:
        return self.matches_any(name, KEY_PATTERNS)

    def get_main_language(self) -> str:
        code_categories = [
            "typescript",
            "javascript",
            "python",
            "csharp",
            "java",
            "go",
            "rust",
            "ruby",
            "php",
        ]
        max_category = "unknown"
        max_count = 0
        for cat in code_categories:
            count = int(self.stats["by_category"].get(cat, 0))
            if count > max_count:
                max_count = count
                max_category = cat
        return max_category

    def describe_top_level_entry(self, name: str, is_dir: bool) -> str:
        if not is_dir:
            lower = name.lower()
            if lower == "readme.md":
                return "Project overview and getting started"
            if lower == "agents.md":
                return "Agent working conventions / repo rules"
            if lower in {"changelog.md", "handoff.md", "contributing.md", "license", "license.md"}:
                return "Project process / legal / handoff documentation"
            if lower in {"package.json", "pnpm-lock.yaml", "yarn.lock", "package-lock.json"}:
                return "JavaScript/TypeScript dependencies and scripts"
            if lower in {"pyproject.toml", "requirements.txt", "poetry.lock", "pipfile", "pipfile.lock"}:
                return "Python dependencies and tooling"
            if lower in {".gitignore", ".gitattributes"}:
                return "Git configuration"
            return "Key file (see tree for context)"

        lower = name.lower()
        if lower == "src":
            return "Main application source code"
        if lower in {"app", "apps"}:
            return "Application(s) (often runnable targets)"
        if lower in {"packages", "pkg"}:
            return "Packages (shared modules/libraries)"
        if lower in {"libs", "lib"}:
            return "Shared libraries/utilities"
        if lower in {"services", "service"}:
            return "Backend services / service layer"
        if lower in {"api", "apis"}:
            return "API layer (routes/controllers/handlers)"
        if lower in {"tests", "test", "__tests__"}:
            return "Tests"
        if lower in {"docs", "documentation"}:
            return "Documentation"
        if lower in {"scripts", "tools"}:
            return "Automation scripts / tooling"
        if lower in {"public", "static", "assets"}:
            return "Static assets"
        if lower == ".github":
            return "GitHub configuration (CI, templates, workflows)"
        if lower == ".claude":
            return "AI tooling / agent workflows (project-local)"
        if lower == ".codex":
            return "Codex CLI skills / configuration (project-local)"
        if lower in {"infra", "infrastructure", "deploy", "ops", "k8s", "terraform"}:
            return "Infrastructure / deployment"
        return "Project-specific directory (see tree)"

    def get_top_level_map(self) -> List[Tuple[str, str, str]]:
        items: List[Tuple[str, str, str]] = []
        try:
            entries = list(self.root_path.iterdir())
        except PermissionError:
            return items

        entries.sort(key=lambda e: (not e.is_dir(), e.name.lower()))
        for entry in entries:
            rel_path = normalize_rel_path(entry.relative_to(self.root_path))
            if self.should_ignore(entry.name, rel_path):
                continue

            is_dir = entry.is_dir()
            display = f"{entry.name}/" if is_dir else entry.name
            typ = "dir" if is_dir else "file"
            purpose = self.describe_top_level_entry(entry.name, is_dir=is_dir)
            items.append((display, typ, purpose))
        return items

    def scan_directory(self, dir_path: Path, depth: int = 0, prefix: str = "") -> List[str]:
        lines: List[str] = []
        if depth > self.max_depth:
            return lines

        try:
            entries = list(dir_path.iterdir())
        except (PermissionError, OSError):
            return lines

        # Sort: directories first, then files
        entries.sort(key=lambda e: (not e.is_dir(), e.name.lower()))

        # Filter ignored
        filtered: List[Path] = []
        for entry in entries:
            try:
                rel = normalize_rel_path(entry.relative_to(self.root_path))
            except ValueError:
                rel = entry.name
            if not self.should_ignore(entry.name, rel):
                filtered.append(entry)

        for idx, entry in enumerate(filtered):
            is_last = idx == len(filtered) - 1
            connector = "`-- " if is_last else "|-- "
            child_prefix = "    " if is_last else "|   "

            try:
                rel_path = normalize_rel_path(entry.relative_to(self.root_path))
            except ValueError:
                rel_path = entry.name

            if entry.is_dir():
                self.stats["total_dirs"] += 1
                lines.append(f"{prefix}{connector}{entry.name}/")
                lines.extend(self.scan_directory(entry, depth + 1, prefix + child_prefix))
                continue

            self.stats["total_files"] += 1
            category = self.categorize_file(entry.name)
            self.stats["by_category"][category] = int(self.stats["by_category"].get(category, 0)) + 1

            if self.is_entry_point(entry.name):
                self.stats["entry_points"].append(rel_path)
            if self.is_config_file(entry.name):
                self.stats["config_files"].append(rel_path)
            if self.is_key_file(entry.name):
                self.stats["key_files"].append(rel_path)

            lines.append(f"{prefix}{connector}{entry.name}")

        return lines

    def scan(self) -> Dict[str, object]:
        root_name = self.root_path.name
        lines = [f"{root_name}/"] + self.scan_directory(self.root_path)
        return {"tree": "\n".join(lines), "stats": self.stats}

    def generate_markdown(self) -> str:
        result = self.scan()
        main_lang = self.get_main_language()
        now = datetime.now().strftime("%Y-%m-%d %H:%M")

        top_level = self.get_top_level_map()
        top_level_table = (
            "\n".join(f"| {p} | {t} | {purpose} |" for p, t, purpose in top_level)
            or "| (none) | - | - |"
        )

        entry_points = "\n".join(f"- {f}" for f in self.stats["entry_points"]) or "- (none detected)"
        config_files = "\n".join(f"- {f}" for f in self.stats["config_files"]) or "- (none detected)"
        key_files = "\n".join(f"- {f}" for f in self.stats["key_files"]) or "- (none detected)"

        distribution = "\n".join(
            f"| {cat} | {count} |"
            for cat, count in sorted(self.stats["by_category"].items(), key=lambda x: -int(x[1]))
        )
        if not distribution:
            distribution = "| (none) | 0 |"

        return f"""# Project Structure Index
> Auto-generated by project-index. Last updated: {now}

## Quick Stats
- **Total files**: {self.stats['total_files']}
- **Total directories**: {self.stats['total_dirs']}
- **Main language**: {main_lang}

## Top-Level Map
| Path | Type | Purpose |
|:---|:---:|:---|
{top_level_table}

## Directory Tree
```
{result['tree']}
```

## Entry Points
{entry_points}

## Config Files
{config_files}

## Key Files
{key_files}

## File Distribution
| Category | Count |
|:---|---:|
{distribution}
"""

    def generate_json(self) -> str:
        result = self.scan()
        return json.dumps(
            {
                "generated": datetime.now().isoformat(),
                "root_path": str(self.root_path),
                "max_depth": self.max_depth,
                "respect_gitignore": self.respect_gitignore,
                "stats": self.stats,
                "main_language": self.get_main_language(),
                "top_level_map": [
                    {"path": p, "type": t, "purpose": purpose} for p, t, purpose in self.get_top_level_map()
                ],
                "tree": result["tree"],
            },
            indent=2,
        )


def main() -> None:
    args = sys.argv[1:]
    root_path = args[0] if len(args) > 0 else "."
    max_depth = int(args[1]) if len(args) > 1 and args[1].isdigit() else 4
    fmt = args[2] if len(args) > 2 and not args[2].startswith("--") else "markdown"

    respect_gitignore = "--no-gitignore" not in args

    scanner = ProjectScanner(root_path, max_depth=max_depth, respect_gitignore=respect_gitignore)
    if fmt == "json":
        print(scanner.generate_json())
    else:
        print(scanner.generate_markdown())


if __name__ == "__main__":
    main()
